lib/core/src/server/dev-server.ts

```diff
@@ -174,17 +174,23 @@ const useProgressReporting = async (
         totalModules = total;
       }
     }
+
     if (value === 1) {
-      options.cache.set('modulesCount', totalModules);
+      if (options.cache) {
+        options.cache.set('modulesCount', totalModules);
+      }
+
       if (!progress.message) {
         progress.message = `Completed in ${printDuration(startTime)}.`;
       }
     }
     reportProgress(progress);
   };
 
-  const modulesCount = (await options.cache.get('modulesCount')) || 1000;
-  new ProgressPlugin({ handler, modulesCount }).apply(compiler);
+  if (options.cache) {
+    const modulesCount = (await options.cache.get('modulesCount')) || 1000;
+    new ProgressPlugin({ handler, modulesCount }).apply(compiler);
+  }
```

Comment:
This one was actually fine. We do still want the ProgressPlugin even if you don't have cache, because it'll be used by the prebuilt manager.

Reply:
I actually prefer the optional chaining, it's just that in the situation below it wasn't the best solution.

Reply:
```ts
let modulesCount = 1000;
if (options.cache) {
  modulesCount = await options.cache.get('modulesCount');
}
new ProgressPlugin({ handler, modulesCount }).apply(compiler);
```

how about it!?

Reply:
See other comment. Looks good now 👍